Mptcp@adress

CrapsDorian opened the pull request:
TODO: - put mptcp@ when the "mptcp" option is in the config
            - fix SIGSEGV in "sock_inet_bind_receiver"
            - add mptcpv4 / v6 in "str2sa_range" when mptcp@

CrapsDorian committed:
FEATURE: add opt-in MPTCP support

Multipath TCP (MPTCP), standardized in RFC8684 [1], is a TCP extension
that enables a TCP connection to use different paths.

Multipath TCP has been used for several use cases. On smartphones, MPTCP
enables seamless handovers between cellular and Wi-Fi networks while
preserving established connections. This use-case is what pushed Apple
to use MPTCP since 2013 in multiple applications [2]. On dual-stack
hosts, Multipath TCP enables the TCP connection to automatically use the
best performing path, either IPv4 or IPv6. If one path fails, MPTCP
automatically uses the other path.

To benefit from MPTCP, both the client and the server have to support
it. Multipath TCP is a backward-compatible TCP extension that is enabled
by default on recent Linux distributions (Debian, Ubuntu, Redhat, ...).
Multipath TCP is included in the Linux kernel since version 5.6 [3]. To
use it on Linux, an application must explicitly enable it when creating
the socket. No need to change anything else in the application.

This attached patch adds an mptcp option in the config which allows the
creation of an MPTCP socket instead of TCP on Linux. If Multipath TCP
is not supported on the system, an error will be reported and the start failed.

Due to the limited impact within a data center environment,
we have decided not to implement MPTCP between the proxy and the servers.
The high-speed, low-latency nature of data center networks reduces
the benefits of MPTCP, making the complexity of its implementation
unnecessary in this context.

Developed with the help of Matthieu Baerts (matttbe@kernel.org) and
Olivier Bonaventure (Olivier.Bonaventure@uclouvain.be)

Link: https://www.rfc-editor.org/rfc/rfc8684.html [1]
Link: https://www.tessares.net/apples-mptcp-story-so-far/ [2]
Link: https://www.mptcp.dev [3]

CrapsDorian committed:
FEATURE: add opt-in MPTCP support

Multipath TCP (MPTCP), standardized in RFC8684 [1], is a TCP extension
that enables a TCP connection to use different paths.

Multipath TCP has been used for several use cases. On smartphones, MPTCP
enables seamless handovers between cellular and Wi-Fi networks while
preserving established connections. This use-case is what pushed Apple
to use MPTCP since 2013 in multiple applications [2]. On dual-stack
hosts, Multipath TCP enables the TCP connection to automatically use the
best performing path, either IPv4 or IPv6. If one path fails, MPTCP
automatically uses the other path.

To benefit from MPTCP, both the client and the server have to support
it. Multipath TCP is a backward-compatible TCP extension that is enabled
by default on recent Linux distributions (Debian, Ubuntu, Redhat, ...).
Multipath TCP is included in the Linux kernel since version 5.6 [3]. To
use it on Linux, an application must explicitly enable it when creating
the socket. No need to change anything else in the application.

This attached patch adds an mptcp option in the config which allows the
creation of an MPTCP socket instead of TCP on Linux. If Multipath TCP
is not supported on the system, an error will be reported and the start failed.

Due to the limited impact within a data center environment,
we have decided not to implement MPTCP between the proxy and the servers.
The high-speed, low-latency nature of data center networks reduces
the benefits of MPTCP, making the complexity of its implementation
unnecessary in this context.

Developed with the help of Matthieu Baerts (matttbe@kernel.org) and
Olivier Bonaventure (Olivier.Bonaventure@uclouvain.be)

Link: https://www.rfc-editor.org/rfc/rfc8684.html [1]
Link: https://www.tessares.net/apples-mptcp-story-so-far/ [2]
Link: https://www.mptcp.dev [3]

CrapsDorian committed:
FEATURE: add opt-in MPTCP support

Multipath TCP (MPTCP), standardized in RFC8684 [1], is a TCP extension
that enables a TCP connection to use different paths.

Multipath TCP has been used for several use cases. On smartphones, MPTCP
enables seamless handovers between cellular and Wi-Fi networks while
preserving established connections. This use-case is what pushed Apple
to use MPTCP since 2013 in multiple applications [2]. On dual-stack
hosts, Multipath TCP enables the TCP connection to automatically use the
best performing path, either IPv4 or IPv6. If one path fails, MPTCP
automatically uses the other path.

To benefit from MPTCP, both the client and the server have to support
it. Multipath TCP is a backward-compatible TCP extension that is enabled
by default on recent Linux distributions (Debian, Ubuntu, Redhat, ...).
Multipath TCP is included in the Linux kernel since version 5.6 [3]. To
use it on Linux, an application must explicitly enable it when creating
the socket. No need to change anything else in the application.

This attached patch adds an mptcp option in the config which allows the
creation of an MPTCP socket instead of TCP on Linux. If Multipath TCP
is not supported on the system, an error will be reported and the start failed.

Due to the limited impact within a data center environment,
we have decided not to implement MPTCP between the proxy and the servers.
The high-speed, low-latency nature of data center networks reduces
the benefits of MPTCP, making the complexity of its implementation
unnecessary in this context.

Developed with the help of Matthieu Baerts (matttbe@kernel.org) and
Olivier Bonaventure (Olivier.Bonaventure@uclouvain.be)

Link: https://www.rfc-editor.org/rfc/rfc8684.html [1]
Link: https://www.tessares.net/apples-mptcp-story-so-far/ [2]
Link: https://www.mptcp.dev [3]

CrapsDorian committed:
Merge branch 'mptcp@adress' of github.com:CrapsDorian/haproxy into mptcp@adress

matttbe commented on src/proto_tcp.c:
```diff
@@ -141,6 +141,41 @@ struct protocol proto_tcpv6 = {
+
+
+// INITCALL1(STG_REGISTER, mptcpv4_protocol_register, &proto_mptcpv4);
+// INITCALL1(STG_REGISTER, mptcpv6_protocol_register, &proto_mptcpv6);
```
detail: here, you can pass `NULL` instead of `&proto_mptcpv4` if you don't use `proto` above.

(but I guess you want this to be not commented)

matttbe commented on src/tools.c:
```diff
@@ -1350,7 +1375,11 @@ struct sockaddr_storage *str2sa_range(const char *str, int *port, int *low, int
+	if (mptcp){
+		printf(" avec mptcp \n");
+		// new_proto = &proto_mptcpv4;
+	}
```
you don't need this, but instead, you can change the code below assigning `new_proto` with:

```c
		new_proto = protocol_lookup(ss.ss_family,
					    proto_type,
					    ctrl_type == SOCK_DGRAM || !!mptcp);
```

matttbe commented on src/cfgparse.c:
```diff
@@ -156,10 +173,12 @@ int str2listener(char *str, struct proxy *curproxy, struct bind_conf *bind_conf,
 
-		ss2 = str2sa_range(str, NULL, &port, &end, &fd, &proto, NULL, err,
+	
+		ss2 = str2sa_range(tcp4_prefix(str), NULL, &port, &end, &fd, &proto, NULL, err,
```
can you remove `tcp4_prefix()`: I'm not sure to understand what you want to do (and will cause memleak)

CrapsDorian committed:
FEATURE: try to add mptcp@

TODO: - put mptcp@ when the "mptcp" option is in the config
             - fix SIGSEGV in "sock_inet_bind_receiver"
             - add mptcpv4 / v6 in "str2sa_range" when mptcp@
